Forward UTM params to petition service

alexsapps opened the pull request.

alexsapps committed:
Forward UTM params to petition service

jakehobbs commented on pages/_app.tsx:
```diff
@@ -18,6 +19,10 @@ const DONATE_MODAL_END_DATE = parseISO("2022-05-31T17:00:00Z");
 function MyApp({ Component, pageProps }: AppProps) {
   const [isDonateModalOpen, setDonateModalOpen] = useState(false);
 
+  useEffect(() => {
```
❓ ❗`useEffect` without any deps will run on every render. is that intentional?

alexsapps replied:
ugh omg ty for catching that. i think i deleted this and tried to put it back from memory but forgot to put back the empty dependency array.

jakehobbs commented on components/HomePage/Letter.tsx:
```diff
@@ -319,3 +322,24 @@ export const LetterForm = () => {
+
+function getUtmParams() {
+  const utmParams : {[key: string]: string} = {};
+  const sessionUtm = adReferralService.getSessionUtmParams();
```
🌈 it's pretty confusing to me that we're referring to these as "session" things everywhere, but actually using Local Storage, not Session Storage.

jakehobbs commented on components/HomePage/Letter.tsx:
```diff
@@ -319,3 +322,24 @@ export const LetterForm = () => {
+    }
+    if (sessionUtm.content !== null) {
+      utmParams["utm_content"] = sessionUtm.content;
+    }
```
🌈 i feel like this would be a lot simpler if we just name the properties w/ the `utm_` prefix within the utm service. any reason for not doing that?

alexsapps replied:
doing so would tie this model to the petition service API which would not be my intent. e.g. if the petition service renamed the args to be camelcase, we shouldn't have to change the session service or have to redesign the connection, imo. and i like that this makes extra clear there's a translation from model to API happening.

jakehobbs commented on services/AdReferralService.ts:
```diff
@@ -0,0 +1,163 @@
+ * AdReferralService persists UTM parametes in Local Storage and allows for
+ * later retrieval of those parameters up until the data expires.
+ */
+class AdReferralService {
```
🌈 this whole thing feels kinda over-engineered and not react-y to me, but i won't block on it. (also maybe that opinion is just due to opinions of the people i normally work with. 😆)

i prob would have just made a hook like this (super quick untested napkin code):

```ts
// zod (already installed in this app) makes parsing json at runtime super easy and 100% typ-esafe
const ReferralParams = z.object({
	utm_source: z.string().nullable().default(null),
	utm_medium: z.string().nullable().default(null),
	utm_campaign: z.string().nullable().default(null),
	utm_content: z.string().nullable().default(null),
})

const ReferralParamsStorage = z.object({
    data: ReferralParams,
	createdAt: z.number(),
})

const useReferralParams = () => {
	// next.js has a useRouter hook so you don't need to be directly interacting w/ the window
	const router = useRouter()
    // reads type-safe query params from url
	const queryParams = useMemo(() => {
		const parsed = ReferralParams.safeParse(router.query)
		if (!parsed.success || !some(Object.values(parsed.data))) {
			return null
		}
		return parsed.data
	}, [router.query])

	// i recommend useLocalStorage from https://github.com/juliencrn/usehooks-ts
	const [getLocalStorage, setLocalStorage, clearLocalStorage] = useLocalStorage(KEY)

    // persists query params to localstorage
	useEffect(() => {
		if (!queryParams) {
			return
		}
		setLocalStorage({ data: queryParams, createdAt: Date.now() })
	}, [queryParams])

   // reads query params back out from localstorage & parses & makes sure they aren't too old
	const getValue = useCallback(() => {
		const val = getLocalStorage()
		if (!val) {
			return null
		}
		const parsed = ReferralParamsStorage.safeParse(val)
		if (!parsed.success) {
			return null
		}
		const {createdAt, data} = parsed.data
		if (new Date(createdAt) < sub(neq Date(), {days: 30})) { //`sub` is from date-fns (already installed)
			return null
		}
		return data
	}, [getLocalStorage])

	return getValue
}
```

alexsapps replied:
perhaps i should use Zod for serialization next time.  but i have a feeling we'll want to reuse this code in other apps which may not be Angular apps (in which case fewer dependencies is better incl Zod and React), and there is a good amount of functionality that i think is easier to test when it's Angular-agnostic.

alexsapps committed:
AdReferralService bug and style fixes:
* Only not not initialize service on each rerender of the app
* Only use "window" in code that is guaranteed to run in the browser and
  not server side rendering context.
* Avoid use of "session" in names to avoid confusion with browser
  sessions.

TESTED by commenting out call to /message/create locally and manually
testing call to production petition service.
